src/Compilers/VisualBasic/Portable/Lowering/LocalRewriter/LocalRewriter_Query.vb

```diff
@@ -162,16 +149,57 @@ Namespace Microsoft.CodeAnalysis.VisualBasic
                 If propertyDefName.StartsWith("$"c, StringComparison.Ordinal) AndAlso Not String.Equals(propertyDefName, StringConstants.Group, StringComparison.Ordinal) Then
                     ' Nested compound variable.
                     Debug.Assert(propertyDefName.Equals(StringConstants.It) OrElse propertyDefName.Equals(StringConstants.It1) OrElse propertyDefName.Equals(StringConstants.It2))
-                    PopulateRangeVariableMapForAnonymousType(syntax, getCallOrPropertyAccess, rangeVariables, firstUnmappedRangeVariable)
-
+                    PopulateRangeVariableMapForAnonymousType(syntax, getCallOrPropertyAccess, rangeVariables, firstUnmappedRangeVariable, rangeVariableMap, inExpressionLambda)
                 Else
                     Debug.Assert(IdentifierComparison.Equals(propertyDefName, rangeVariables(firstUnmappedRangeVariable).Name))
-                    _rangeVariableMap.Add(rangeVariables(firstUnmappedRangeVariable), getCallOrPropertyAccess)
+                    rangeVariableMap.Add(rangeVariables(firstUnmappedRangeVariable), getCallOrPropertyAccess)
                     firstUnmappedRangeVariable += 1
                 End If
             Next
         End Sub
 
+        Friend Shared Function RewriteQueryLambdaBody(
+            expressionRewriter As Func(Of BoundExpression, BoundExpression),
+            originalNode As BoundQueryLambda,
+            rangeVariableMap As Dictionary(Of RangeVariableSymbol, BoundExpression),
+            Optional instrumenterOpt As Instrumenter = Nothing,
+            Optional instrumentQueryLambdaBody As Boolean = False) As BoundStatement
+
+            Dim returnstmt As BoundStatement = New BoundReturnStatement(originalNode.Syntax,
+                                                                        expressionRewriter(originalNode.Expression),
+                                                                        Nothing,
+                                                                        Nothing)
+
+            If instrumentQueryLambdaBody Then
+                returnstmt = instrumenterOpt.InstrumentQueryLambdaBody(originalNode, returnstmt)
```

Comment:
>returnstmt = instrumenterOpt.InstrumentQueryLambdaBody(originalNode, returnstmt) [](start = 16, length = 80)

This can be done in the caller, then we wouldn't have to pass around information that makes sense only in the LocalRewriter.

Reply:
Sure, will refactor.

---
In reply to: [137420732](https://github.com/dotnet/roslyn/pull/21356#discussion_r137420732) [](ancestors = 137420732)